src/ic-cdk/src/api/stable.rs

```diff
@@ -147,28 +164,42 @@ impl io::Write for StableWriter {
 ///
 /// Keeps an offset and reads off stable memory consecutively.
 pub struct StableReader {
-    /// The offset of the next write.
+    /// The offset of the next read.
     offset: usize,
+    /// The capacity, in pages.
+    capacity: u32,
 }
 
 impl Default for StableReader {
     fn default() -> Self {
-        Self { offset: 0 }
+        Self {
+            offset: 0,
+            capacity: stable_size(),
+        }
     }
 }
 
 impl StableReader {
     /// Reads data from the stable memory location specified by an offset.
     pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, StableMemoryError> {
-        stable_read(self.offset as u32, buf);
-        self.offset += buf.len();
-        Ok(buf.len())
+        let cap = (self.capacity as usize) << 16;
```

Comment:
I have a version of these for 64-bit memory where I did something similar to prevent out-of-bounds access.

In my version, I didn't introduce a `self.capacity` field because when the memory is grown it became overly complicated to keep in sync.

For example, using the writer may result in growing the memory. In this version,  `self.capacity` would not be updated and become out of sync.

Instead, I did the equivalent of `let cap = (stable_size() as size) << 16;` here.